### PR TITLE
ROI.py, getArrayRegion: Fix return mapped coordinates

### PR DESCRIPTION
**Pull request by malikolivier**
The *getArrayRegion* method is defined as returning a tuple of the points
in the selected region and the mapped coordinates if the
*returnMappedCoords* keyword argument is set to True in the parent class
*ROI*.

In the *EllipseROI* and *PolyLineROI* classes, *getArrayRegion* was overriden,
however it ignores the *returnMappedCoords* keyword argument, leading to
unintended bugs because of the change in interface between the parent class and
the subclass.

This patch fixes the above bug.
If *returnMappedCoords* is set to False, then only *arr* containing the
array region is returned. If *returnMappedCoords* is set to True, a
tuple of the array region and the mapped coordinates is returned.

EDIT: Fixed the same bug for PolyLineROI.

@campagnola Do you require unit tests? I wanted to write some but I gave up for now as I cannot get unit tests to succeed in my environment (maybe I missed something in the setup):

```
$ python --version
Python 3.6.3
$ pip --version
pip 9.0.1 from /home/malik/workspace/pyqtgraph/venv/lib/python3.6/site-packages (python 3.6)
$ pip freeze
apipkg==1.4
attrs==17.4.0
coverage==4.4.2
execnet==1.5.0
numpy==1.14.0
pkg-resources==0.0.0
pluggy==0.6.0
py==1.5.2
PyQt5==5.9.2
pytest==3.3.2
pytest-cov==2.5.1
pytest-forked==0.2
pytest-xdist==1.22.0
scipy==1.0.0
sip==4.19.6
six==1.11.0
$ py.test
7 failed, 24 passed, 129 skipped in 4.04 seconds
```

### CODE REVIEW

File: pyqtgraph/graphicsItems/ROI.py [hunk 1]
```diff
@@ -1653,7 +1662,10 @@ def getArrayRegion(self, arr, img=None, axes=(0, 1), **kwds):
         shape = [(n if i in axes else 1) for i,n in enumerate(arr.shape)]
         mask = mask.reshape(shape)
         
-        return arr * mask
+        if returnMappedCoords:
+            return arr * mask, mappedCoords * mask
```
Comment (campagnola) [inline pyqtgraph/graphicsItems/ROI.py]: Doesn't make sense to me to mask the coordinates as well--each element in the returned array has a corresponding location in the original data, even if that element value was masked.

Comment (malikolivier) [reply pyqtgraph/graphicsItems/ROI.py]: You're right. I pushed a fix. Let me know your thoughts.

**Comment by campagnola** [comment]
Try running `examples/ROIExamples.py` -- these changes break the triangular ROI in the upper-left panel.

File: pyqtgraph/graphicsItems/ROI.py [hunk 2]
```diff
@@ -1914,15 +1926,23 @@ def shape(self):
...
         """
         br = self.boundingRect()
         if br.width() > 1000:
             raise Exception()
-        sliced = ROI.getArrayRegion(self, data, img, axes=axes, fromBoundingRect=True, **kwds)
+        if returnMappedCoords:
+            sliced, mappedCoords = ROI.getArrayRegion(self, data, img,
+                                                      axes=axes,
+                                                      fromBoundingRect=True,
+                                                      **kwds)
```
Comment (campagnola) [inline pyqtgraph/graphicsItems/ROI.py]: do you need `returnMappedCoords=True` here?

Comment (malikolivier) [reply pyqtgraph/graphicsItems/ROI.py]: You are right. This is utterly wrong. I got confused and I mixed up the use of  *fromBoundingRect* and *returnMappedCoords*.

I pushed a fix.

**Comment by malikolivier** [comment]
3bc971e fixes the triangular ROI in the upper-left panel of `examples/ROIExamples.py`.

**Comment by malikolivier** [comment]
Up. Is it Ok to merge?

**Comment by paulmueller** [comment]
likely related to #292

**Comment by j9ac9k** [comment]
Closing and Opening to trigger CI run.

**Comment by j9ac9k** [comment]
CI Failure is unrelated to this PR.

**Comment by j9ac9k** [comment]
Closing/Opening to trigger new CI run....

**Comment by j9ac9k** [comment]
I just checked out this issue, I was thinking it might address #292 but it does not appear to do so...I'm still a little puzzled with the intended behavior so I'm going to investigate a bit more with `ImageAnalysis.py` example.

**Comment by j9ac9k** [comment]
Also CI passed for all pipelines

File: pyqtgraph/graphicsItems/ROI.py (hunk 1 above)
Comment (rwalroth) [reply pyqtgraph/graphicsItems/ROI.py]: I disagree, makes more sense to mask out coordinates. I implemented a different fix that uses a flag value to mark which coordinates were masked in the returned array, this also takes into account that coords is a 3 dimensional array:
```
def maskCoordinates(coords, mask, flag=-1):
    """
    Applies the mask to the coordinates, setting any 0 values in mask
    to -1.

    Parameters
    ----------
    coords : numpy.ndarray, coordinates returned from getArrayRegion
    mask : numpy.ndarray, mask to be applied
    flag : int, value to represent 0 values in mask

    Returns
    -------
    m_coords : numpy.ndarray, masked coordinates

    """
    m_coords = coords.copy()
    m_coords[0] *= mask
    m_coords[1] *= mask
    m_coords[0][mask == 0] = flag
    m_coords[1][mask == 0] = flag

    return m_coords
```

**Comment by outofculture** [comment]
This had merge conflicts with the recent TriangleROI work, which I resolved on [a copy of this branch](https://github.com/outofculture/pyqtgraph/tree/fix-ROI-getArrayRegion). @malikolivier you can merge from my branch, resolve them yourself, or I can reopen this PR against the copy. However you'd prefer.

This branch behaves identically to master with respect to #292, but I'm in @j9ac9k's boat of being confused about that bug.

This branch definitely fixes #1054 

I couldn't see any other negative impact from this code. Ellipses, triangles and poly ROIs all report seemingly accurate values.

**Comment by outofculture** [comment]
Superceded by #1700

**Pull request closed without merging**